Issue 27 implementation

eric239 opened the pull request:
Factored out rendering by accept header and implemented HTML rendering using freemarker

eric239 commented:
The following endpoints support HTML-friendly output with cross-linking as appropriate:
- `/` (e.g. `http://localhost:2876/schema-repo`) 
- `/{subject}/all`
- `/{subject}/latest`
- `/{subject}/id/{id}`
- `/{subject}/config`
- `/config`

FelixGV commented:
Awesome! I will try to review this today. Thanks @eric239 !

eric239 commented:
No rush, if it happens to be later this week that's fine as well.

Side note: as you will see, my presentation (HTML) skills are not best, but
that shouldn't matter in this case :-)

On Tue, Jan 13, 2015 at 9:43 AM, Felix GV <notifications@github.com> wrote:

> Awesome! I will try to review this today. Thanks @eric239
> <https://github.com/eric239> !
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/schema-repo/schema-repo/pull/35#issuecomment-69785976>
> .
>

FelixGV commented on client/src/main/java/org/schemarepo/client/Avro1124RESTRepositoryClient.java:
```diff
@@ -93,7 +93,7 @@ public Subject lookup(String subject) {
     ArrayList<Subject> subjectList = new ArrayList<Subject>();
     try {
-      String subjects = webResource.get(String.class);
+      String subjects = webResource.accept(MediaType.TEXT_PLAIN_TYPE).get(String.class);
```
This worries me a little bit...

The intent of the ```Avro1124RESTRepositoryClient``` class is to have a test which emulates, as much as humanly possible, the behavior of the original client implementation in the AVRO-1124 ticket, so that folks using that code can upgrade to the schema-repo as seamlessly as possible.

Adding this accept clause seems to imply that the default (or unspecified) media type sent by the Jersey client implementation would break against the new schema repo server code. If this is false, then can we just remove the ```accept(MediaType.TEXT_PLAIN_TYPE)``` statement? If this is true, then does that mean this PR introduces server-side changes that are incompatible with the old (AVRO-1124) clients, and if so, is there anything we can do to prevent or mitigate that?

eric239 replied:
(Unfortunately) it happens so that the default `Accept` value is
`text/html`. It's idiosyncrasy of jersey client.

Perhaps RESTRepository needs a new parameter to enable HTML support (so
that whoever needs strict compatibility with the original behavior can turn
it off). Any other thoughts?

On Tue, Jan 13, 2015 at 9:51 AM, Felix GV <notifications@github.com> wrote:

> In
> client/src/main/java/org/schemarepo/client/Avro1124RESTRepositoryClient.java
> <https://github.com/schema-repo/schema-repo/pull/35#discussion-diff-22879824>
> :
>
> > @@ -93,7 +93,7 @@ public Subject lookup(String subject) {
> >    public Iterable<Subject> subjects() {
> >      ArrayList<Subject> subjectList = new ArrayList<Subject>();
> >      try {
> > -      String subjects = webResource.get(String.class);
> > +      String subjects = webResource.accept(MediaType.TEXT_PLAIN_TYPE).get(String.class);
>
> This worries me a little bit...
>
> The intent of the Avro1124RESTRepositoryClient class is to have a test
> which emulates, as much as humanly possible, the behavior of the original
> client implementation in the AVRO-1124 ticket, so that folks using that
> code can upgrade to the schema-repo as seamlessly as possible.
>
> Adding this accept clause seems to imply that the default (or unspecified)
> media type sent by the Jersey client implementation would break against the
> new schema repo server code. If this is false, then can we just remove the
> accept(MediaType.TEXT_PLAIN_TYPE) statement? If this is true, then does
> that mean this PR introduces server-side changes that are incompatible with
> the old (AVRO-1124) clients, and if so, is there anything we can do to
> prevent or mitigate that?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/schema-repo/schema-repo/pull/35/files#r22879824>.
>

FelixGV replied:
Ah... that's a bummer.

Although it hasn't been stated explicitly (yet), I'd like one of the design goals of this project to be easy upgrade-ability for users of the AVRO-1124 code. That being said, I think we should balance that need against the goal of offering as much as possible (ideally all) of the new features while still maintaining that compatibility. Your proposed solution of having a parameter to enable HTML support works for the backwards compatibility goal, but it's too bad that it would essentially make HTML browsing unavailable for anyone who hasn't fully migrated ALL of their clients to the new implementation... don't you think?

I'm not a hardcore RESTafarian so I don't know what the purists would suggest. I imagine they would say it's best to have a single endpoint with multiple supported content types. That being said, being pragmatic, I think the best thing to do in this case would be to have several endpoints...

So perhaps we can have a root path which only supports plain-text (and thus, older clients), and another root path which supports all content types? Ideally, we should aim to support both with the same code base, by simply instantiating multiple ```RESTRepository``` instances with the same ```Repository``` instance passed in but different sets of ```Renderer```s. Or something along those lines (I'm not attached to any specific implementation strategy, but the goal would be to maximizes code re-use and cache things only once).

If making both work at the same time seems like too much work, we can begin with your proposition of making HTML support globally toggled (with the "on" setting breaking old clients), but if we do that, I think the default config should be "off" (unfortunately...), and that we should follow up with a separate issue to enable both simultaneously/gracefully.

What do you think?

eric239 replied:
Ugh... this is getting complicated. Just how widespread do you think the use of old AVRO-1124 client is? Are we jumping through hoops b/c of a few pieces of anecdotal evidence or there's something more substantial? To be clear, I'm not challenging your statement, just want to have a better underlying picture.

Now to the tech side of it... One thing I can do is to look at User-Agent header. Presumably jersey won't have it set at all (or set to a specific value) whereas a real browser will of course always have it. This smells of a kludge but seems to me a "lesser evil". I'll report back once I reconfirm my assumptions regarding UA.

FelixGV replied:
It is hard to gauge the use of AVRO-1124 since it is a back-end component. Here are some indicators, however (as of 2014-01-14) :

* The [Apache JIRA ticket](https://issues.apache.org/jira/browse/AVRO-1124) has:
    * 30 votes, which ranks it:
        * [20th of all time among all Apache projects](https://issues.apache.org/jira/issues/?jql=%20order%20by%20votes%20desc)
        * [4th in currently open tickets among all Apache projects](https://issues.apache.org/jira/issues/?jql=status%20%3D%20OPEN%20order%20by%20votes%20desc) 
        * [1st of all time in AVRO](https://issues.apache.org/jira/issues/?jql=project%20%3D%20AVRO%20order%20by%20votes%20desc)
    * 61 watchers, which ranks it:
        * [36th of all time among all Apache projects](https://issues.apache.org/jira/issues/?jql=order%20by%20watchers%20desc).
        * [11th in currently open tickets among all Apache projects](https://issues.apache.org/jira/issues/?jql=status%20%3D%20open%20order%20by%20watchers%20desc)
        * [1st of all time in AVRO](https://issues.apache.org/jira/issues/?jql=project%20%3D%20AVRO%20order%20by%20watchers%20desc)
    * A flurry of comments, spanning 2.5 years.
* The [Camus](https://github.com/linkedin/camus) project uses it as a component.
    * Camus is (probably) the most popular way to move data from Kafka to Hadoop, having 94 watchers, 377 stars, 237 forks and 26 unique contributors on GitHub.
    * The term "1124" has been [mentioned in 18 different threads on the Camus mailing list](https://groups.google.com/forum/#!searchin/camus_etl/1124). Most of those threads have several people participating and hundreds of views each (according to Google Groups anyway, which presumably includes only web views, not opened emails, since that is never properly traceable).

The schema repo is meant to make evolution of schemas easier. It is all about backwards and forwards compatibility. Thus, any change that requires a "big bang" style of integration (where both clients and server need to be upgraded simultaneously) will severely impact its usability and growth.

Offering different results based on the UA seems brittle, and goes against the value proposition of REST (being language-agnostic)... I don't see much of a problem with offering an HTML view of the repo on a different URL root, however... Having it all under one root is a nice to have, but correctness and compatibility are more important, IMHO.

eric239 replied:
Well... if we have separate HTML endpoints that will essentially duplicate most of the `RESTRepository` methods (with different `@Path` annotations) *or* it will be simpler to have a separate `HTMLRepository` which will be living parallel to `RESTRepository`. Is that what you have in mind?

FelixGV replied:
I think there might be a simpler/cleaner way (just brainstorming here).

It would probably be possible to modify ```RepositoryServer``` so that it binds multiple ```RESTRepository``` instances to different root paths. This way, you could have ```localhost:2876/schema-repo``` serving just the machine-readable Content-Types (as it does currently, without this PR) and ```localhost:2876/schema-repo-browser``` serving the human-readable content. Both ```RESTRepository``` instances could be sharing 100% the same code, and be passed in the same ```Repository``` instance so that caching, state and synchronization is fully shared, but a different set of ```Renderer``` so that they can respond (and reject) requests appropriately.

I'm not sure exactly how to do this in Jersey, but there's got to be a way... It seems like a very typical use case.

Thoughts?

FelixGV commented on server/src/main/java/org/schemarepo/server/JsonRenderer.java:
```diff
@@ -0,0 +1,70 @@
+
+  @Override
+  public String renderProperties(Properties props, String comment) {
+    // TODO: when the need arises, this may yield an actual JSON
```
Perhaps we should throw an exception instead of advertising support for something we don't implement? The server could then return ```415 Unsupported Media Type``` if that happens. Thoughts?

Of course, implementing the actual ```renderProperties``` as json would be fairly trivial, but I think it makes sense to use 415s wherever appropriate. For example, let's say someone down the line wants to contribute back an XML renderer but is missing one of the interface's functions, it should be our convention to return a proper error code rather than falsely advertise support for a certain content type.

My two pesos.

FelixGV replied:
Or perhaps a more relevant example might be: let's say you decide to add an HTML response for one of the other APIs, it would be desirable for you to be able to do that without the inertia of needing to implement the corresponding plain text and json APIs as well. If our client can gracefully deal with this, then I think we'll be better off in the future, by being able to move more quickly and reliably.

eric239 replied:
I'm not sure I understand what do you propose.

In order for an endpoint to serve different response depending on the `Accept` header, the corresponding functionality must be provided in the `Renderer` interface. The alternative is to bake the logic ("if mediaType is ...") into `RESTRepository` itself, which I find less attractive stylistically and as far as separation of concerns goes.

In this particular case, I'm saying (through code) that we don't elect to render something in a particular way yet. This is consistent / compatible with the original behavior of completely ignoring `Accept` header. Responding with `415` may be more correct formally, but I don't see why would we want to do that from practical point of view.

I see your concern of "converting" an endpoint to support different views based on the media type as being a little heavy (respective implementations in all classes). I contemplated going down the path of not having a method per endpoint but rather having a generic `render` taking a map of params and an enum for an operation/endpoint, but I thought that would make the code unnecessarily more complex. We have a very limited number of endpoints, most of which already support per-content views, and I don't see how that can grow significantly. Regarding having HTML-only endpoints, I don't quite see how does that fit into existing realm. My goal was to provide quick (and perhaps somewhat dirty) way of accessing the repo without access to underlying backend or having to use the likes of `curl`. It has never involved creating a full-blown website on top of the repo.

FelixGV replied:
I guess there are a few aspects to this...

First of all, there are two notions of "default":

1. What the default response should be when no Accept header at all is specified. Here, I think we agree plain text is fine...
2. What the default response should be when an unsupported Accept header is specified. This is where we disagree, I think. I think it is safer to fail explicitly (via 415) than to make assumptions on a client's ability to process plain text (or some other default) when they explicitly asked for something else.

Secondly, what component should choose how to behave when a ```Renderer``` doesn't support a given operation? I believe it should be the ```RESTRepository``` that chooses what to do, not the ```Renderer```. Otherwise, there is code duplication and coupling of the ```Renderer``` with what should be a default and generic error handling code. If we decide to change what the function call should be for handling the plain text response, that should not require modifying the corresponding unimplemented operations in unrelated renderers....

Finally, regarding HTML-only endpoints, I didn't really mean that we should have that. I just meant that if we want to implement an HTML version of an endpoint that is currently just in plain text, you would need to add that operation to the ```Renderer``` interface, and it would remain unimplemented in the json one, thus further expanding the gap in functionality between the various renderers.

eric239 replied:
> What the default response should be when an unsupported Accept header is specified.

That's where I miss your point. There's currently (in the master) no such thing as the above. Do you intend to deviate from that? Wouldn't that be an additional backward comp. break?

FelixGV replied:
Thanks, you bring up a very good point @eric239.

I did a bit of digging and here is what I found out:

In master:

* ```415 Unsupported Media Type``` responses are only used for write operations where the ```Content-Type``` header is inappropriate. I think this generally makes sense (though we may want to be a bit more lax in certain cases, as requested in Issue #3, but let's ignore that for the purpose of this discussion).
* Read operations apparently "never fail", no matter what the ```Accept``` header is. This second point surprised me because I vaguely recalled getting some errors in the past. It does mean your previous comment is right, however.

Before the current master:

It turns out that I changed the behavior in 4932f5494d29d461cab746b4667b728f06cccb5e (as part of the work on Issue #17 to support multi-line schemas using JSON responses). Before then, the server would return ```406 Not Acceptable``` when an unsupported ```Accept``` header was requested. My memory failed me in that it wasn't previously returning ```415```, but I think this old behavior (```406```) was more proper than the current one (not honoring the request's ```Accept``` header and making assumptions). I should have been more careful when working on the multi-line issue.

I believe the old behavior happened because the ```RESTRepository``` was prefixed by the ```@Produces(MediaType.TEXT_PLAIN)``` annotation, which I then removed as part of 4932f5494d29d461cab746b4667b728f06cccb5e. By the way, when running 4932f5494d29d461cab746b4667b728f06cccb5e, doing a request with ```Accept: text/html``` also returns a ```406``` error code, so I'm not sure what's the deal with what you were saying about the Jersey client always requesting text/html? (I haven't looked into verifying that...)

I would like it if we could bring back the old behavior, though I'm not sure if it should be done by resuming to use ```@Produces``` annotations (which are completely eliminated in the current master), or if we should somehow manually throw 406 errors. At first glance, I haven't seen a Jersey exception which represents 406, so I don't know if there is any way of programmatically triggering those, or if the ```@Produces``` annotations are the only way...

This is a bit of a mess, and my fault. I'm willing to put some time in fixing it (unless you want to), but as of now I'm not sure what the best solution would be. I would need to read up and tinker some more...

Thoughts?

eric239 replied:
If we could confirm that the original behavior was "strict" with respect to `Accept` header (responding with `406` when it wasn't plain text), that would be an easy way out of this tangle - I can put logic in `RESTRepository` to strictly comply with the header. I have reconfirmed that current client sends `Accept: text/html, image/gif, image/jpeg, *; q=.2, */*; q=.2` when not explicitly instructed, but as you wrote that's probably b/c originally the entire client was annotated with `@Produces(TEXT_PLAIN)`.

I don't think there won't be any issues coding the above as long as we agree that's the plan.

FelixGV replied:
My previous comment was a bit misleading. I *tested* 4932f5494d29d461cab746b4667b728f06cccb5e, because it is the commit prior to 4bb15507073139ed00ed8c24946c94754ae243fe where the changes to ```Accept``` handling actually happened.

So I checked out 4932f5494d29d461cab746b4667b728f06cccb5e, ran it and curled it with ```Accept: text/html``` and I did get a ```406```. I can post the bash commands to reproduce, if you want.

I think the reason why the Jersey REST client worked without explicitly specifying any Accept header is because its default Accept clause includes ```*/*``` (albeit at a lower priority than ```text/html```, which is what is screwing us up at the moment...). Previously, it was the REST *server* that was annotated with ```@Produces(TEXT_PLAIN)```, not the REST client.

I'm sorry for my confusion, but I'm not sure what exactly do you propose to be the plan?

I'm still under the impression that the only sane way we can deal with this is to have ```/schema-repo``` serving machine-readable (and backwards-compatible) content while ```/schema-repo-browser``` serves the human-readable content...

eric239 replied:
Finally getting back to this... Let's try to be real clear about the requirements:

* `/schema-repo`
  * supports `text/plain` and `application/json`, with the former designated as the default
  * responds with with `406` when `Accept` does not include either one of configured media types (as per the above) or `*/*`
  * tries to match configured types (in the order they were registered, e.g. deterministic) against the request, if no match uses the default
* `/scheama-repo-browser`
  * configured with a single media type support `text/hrml` which is also the default
  * thus based on the logic described above will respond with `406` unless `Accept` includes `text/html` or `*/*`

How does that sound? I believe I can shuffle the code around to make that happen...

FelixGV replied:
Hi @eric239,

Thanks for continuing to push on this.

I think both ```/schema-repo``` and ```/schema-repo-browser``` should have the same behavior of responding with ```406``` if there is an explicitly specified set content types in the ```Accept``` header and none of them match with the configured media types. In other words:

* If ```Accept``` is not specified, then it uses the first configured media type as default.
* If ```Accept``` is specified, but the requested end point does not support that content type within its configured media types, then it returns ```406```.

Your proposed requirements seem pretty good, but I think the last part of bullet 1.3 ("if no match uses the default") contradicts bullet 1.2...

eric239 replied:
Okay, I think we're converging...
> If `Accept` is not specified, then it uses the first configured media type as default.

I would modify this to read
* If `Accept` is not specified, or equals to one of `*, */*`, then it uses the first configured media type as default.

If you agree, I'm going to start by writing unit tests to reflect these rules...

FelixGV replied:
Yes, I think your latest re-phrase is good!

FelixGV replied:
And unit tests for this would definitely be great.

FelixGV commented on server/src/main/java/org/schemarepo/server/RESTRepository.java:
```diff
@@ -82,81 +85,66 @@
-
-    }
+    Renderer renderer = getRenderer(mediaType);
+    return Response.ok(renderer.renderSubjects(repo.subjects()), renderer.getMediaType()).build();
```
Looks good.

FelixGV commented on server/src/main/java/org/schemarepo/server/RESTRepository.java:
```diff
@@ -82,81 +85,66 @@
-      throw new RuntimeException(e);
-    }
-    return writer.toString();
+    return getRenderer(mediaType).renderProperties(props, "Configuration of subject " + subject);
```
Hmm... why do you sometimes write the reponse like this and sometimes like that:
```
Renderer renderer = getRenderer(mediaType);
return Response.ok(renderer.renderSubjects(repo.subjects()), renderer.getMediaType()).build();
```
The latter specifies a status code and a content type, but the former does not. Wouldn't it be more proper to always do the former...? Although that raises the question of whether you can deal with errors at all if you're wrapping the response inside a ```Response.ok()```...

Thoughts?

eric239 replied:
The latter syntax involves explicitly setting content type of the response, that's the main (and really the only) difference. (Should there be an exception in processing, it will still result in a 500 status). I'm using the former syntax when the underlying renderer implementations don't actually respond with the appropriately formatted content (e.g. it's not really valid JSON) and thus we don't insist on the response media type and let the client interpret as it wishes. If all the TODOs were implemented, then everything would use the "latter" syntax.

FelixGV commented on server/src/main/java/org/schemarepo/server/RESTRepository.java:
```diff
@@ -370,4 +351,15 @@ private SchemaEntry exists(SchemaEntry entry) {
+    String primaryMediaType = mediaType == null ? null : mediaType.split(",")[0];
+    Renderer r = rendererByMediaType.get(primaryMediaType);
+    if (r == null) {
+      logger.info("No renderer configured for media type {}, defaulting to the renderer for {}", primaryMediaType, DEFAULT_MEDIA_TYPE);
```
I think by default we should return a 415 status code, rather than making assumptions on what content type the client wants. Thoughts?

eric239 replied:
I think `415` would be too strict. Again, in tune of the original behavior, most of the time it's reasonable to assume that the client will be able to deal with the default-rendered (aka plain text) response. This is somewhat philosophical though.

FelixGV commented:
I think these are all the comments I have...

The comment on using 415s should be easy to address, if you agree. The one on backwards-compatibility is probably a little more tricky.

I tested manually, and it's definitely great to have browsing capability!! I can't wait for us to polish the last details so this is merged in (: !

I might want to add some extra features to the UI after this is merged in (breadcrumbs, registration form...).

Cool stuff!

-F

eric239 commented:
Okay. I cannot find a not too crazy way to support `/schema-repo` and `/schema-repo-browser` - there is only one servlet context per jetty app and the path (`/schema-repo`) is set at that "singleton" level. *But* I can make `/schema-repo` and `/schema-repo/browser`. Of course this conflicts with browser as a potential subject name, but if we choose it carefully (e.g. perhaps `/schema-repo/.html`, note the dot) I hope it won't be a real issue. What do you think?

FelixGV commented:
Hmm... This is annoying.

I've done a bit of reading, and it seems using ```GuiceContainer``` is a bit inflexible. It should be possible to serve different paths with different instances of a servlet, [as shown here](https://github.com/google/guice/blob/master/extensions/servlet/src/com/google/inject/servlet/ServletModule.java#L110), but since we serve everything with ```GuiceContainer```, it seems this capability is obscured.

If we want to avoid doing big code/wiring changes (i.e.: if we want to keep using the ```GuiceContainer```), one way to achieve serving ```/schema-repo``` and ```/schema-repo-browser``` might be to configure our context path (currently set via ```Config.JETTY_PATH```) to be ```"/"``` (instead of ```/schema-repo```), and then add an extra prefix within the various paths in ```RESTRepository``` (i.e.: ```@Path("{subject}/latest")``` would become ```@Path("{context}/{subject}/latest")``` and so on). If we did that, then we could use both the ```{context}``` and the ```Accept``` header in order to choose which Renderer to use (HTML or otherwise). This would also ensure full backwards compatibility and no name-clash.

Overloading the topic names seems messy and a bad habit to get into, however unlikely the name collision may be. By the way, the new end-points we added recently, ```/status``` and ```/config```, can also clash with topic names. Moving those up to "context-level" would be desirable as well (i.e.:, they would remain defined as ```@Path("/config")```, NOT changed to ```@Path("{context}/config")```). This would mean we'd have 4 top-level paths we support:

* ```/schema-repo/{subject}```
* ```/schema-repo-browser/{subject}```
* ```/config```
* ```/status```

Settling for this kind of URL scheme will also us to evolve more quickly and reliably in the future. Say, if we want to support another general (not subject-specific) end point, we would be able to add it at the top level, with no concern for name clashes.

I'm sorry for the nit-picking, but I think we need to be super careful about our end point definition and backwards compatibility. Hope you don't mind.

Please let me know what you think.

-F

eric239 commented:
Please take a look now. I've implemented multiple entry points (`schema-repo` and `schema-repo-browser`). Note that
* `/config` and `/status` are left as-is in this commit
* it is unclear how to make the entry points configurable (e.g. use JETTY_PATH property) but it is also unclear (to me) why does this need to be configurable
* for the same reason there are two trivial subclasses of `RESTRepository` - `MachineOrientedRESTRepository` and `HumanOrientedRESTRepository` - that way I can apply different class-level `@Path` annotations and reuse all the code.

FelixGV commented:
Great!

So we can bind the two ```RESTRepository``` extensions with different ```@Path``` annotation values. Works for me. I don't mind them being trivial.

Regarding ```/config``` and ```/status```, I feel bad for overlooking this previously when it was committed, but I think we really need to avoid name-clashes at all cost. Can we extract those two endpoints into another class (or two, if you prefer) and bind it/them like the rest? Perhaps we should make them ```/schema-repo-config``` and ```/schema-repo-status``` to make them a bit more consistent with the rest... But the short version can be fine too.

I'm ok with removing ```@Config.JETTY_PATH```. If someone needs to make the path prefix configurable, they would likely need to make the whole prefix configurable (i.e.: including the ```schema-repo``` part which is now hard-coded in the ```@Path``` annotations), so there's probably little use in exposing configurability for just the context path at the moment. If someone needs it in the future, we can always open an issue and re-visit/re-factor.

It's good to have a unit test. I think it should probably be split it into two: a positive test for the supported content types and a negative one to test that we get ```406``` responses when explicitly asking for an unsupported content type, like we were discussing above. The latter can be returned with ```Response.status(Status.NOT_ACCEPTABLE);```...

I can help with some of this, if you want. Just let me know.

Thanks @eric239 (:

-F

eric239 commented:
I'll finish ​tomorrow evening. I didn't disagree with `/config` and
`/status` to be pulled out, just wanted to let you take a peek at the bulk
of the changes first. Thanks for checking.

FelixGV commented:
Ok, cool (:

Another thing: I just tested the browser part, and the relative links only work on ```http://localhost:2876/schema-repo-browser/``` (with trailing slash) but not on ```http://localhost:2876/schema-repo-browser```. I have no experience with freemarker so I don't know if there's framework-level support for making sure we generate proper URLs (akin to [Play Reverse Routing](https://www.playframework.com/documentation/2.1.1/ScalaRouting#Reverse-routing), for example), but anyway...

This is going to be great to merge in. Thanks again @eric239 (:

-F

eric239 commented:
Got swamped at work, nevertheless expect to finish this week.

eric239 commented:
@FelixGV please review the latest... I believe I've addressed all items mentioned earlier, specifically
* `/config` and `/status` are now root resources
* added more tests around media types handling
* `/schema-repo-browser` redirects to `/schema-repo-browser/`
* some cleanup due to all the changes

FelixGV commented on server/src/test/java/org/schemarepo/server/TestRESTRepository.java:
```diff
@@ -81,19 +86,52 @@ public void testCreateNullSubject() {
+    final String contentType = "Content-Type";
+    repo.createSubject("dummy", new MultivaluedMapImpl());
+    // null and all-inclusive (* or */*) mediaTypes result in the default configured renderer being used
+    for (String mediaType: new String[] {null, "", "*/*", "text/html, image/gif, image/jpeg, *; q=.2, */*; q=.2"}) {
```
Nitpick: maybe we should add ```text/plain``` to the list of tested media types as well?

FelixGV commented on server/src/main/java/org/schemarepo/server/JsonRenderer.java:
```diff
@@ -0,0 +1,70 @@
+
+  @Override
+  public String renderSchemaEntry(SchemaEntry schemaEntry, boolean requestForLatest) {
+    return requestForLatest ? schemaEntry.toString() : schemaEntry.getSchema();
```
If we don't support JSON rendering for this one, can we throw an exception that will bubble up as a ```406``` instead?

FelixGV commented:
Hi @eric239, this looks great!

I have left two very minor comments, which I believe should be very easy to fix.

Can we squash all commits into one and rebase? Please mention Issue #27 in the commit message so we have traceability, as this is a fairly big commit. You may also want to mention that this includes fixes and tests for media type handling.

Thanks a lot!

-F

eric239 commented:
Made the fixes, thanks for pointing out.

Will squash commits later tonight.

eric239 commented:
Ugh... some odd line-ending issues between diff and patch on my home win box, will squash tomorrow when I get to my work mac.

FelixGV commented:
Okay, cool (: ! Thanks again @eric239

FelixGV commented:
Your fix is merged into master (:

Reference:
* Main issue: #27 
* Work in progress PR: #35 
* Final PR: #40 

Thanks a lot @eric239 !
